src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -1432,41 +1464,175 @@ int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl
+      block.deleteMarker = true;
+
+      if (source->get_bucket()->versioned() && !source->get_bucket()->versioning_enabled()) { // if versioning is suspended
+        block.version = "null"; 
+      } else {
+	// create a delete marker
+	enum { OBJ_INSTANCE_LEN = 32 };
+	char buf[OBJ_INSTANCE_LEN + 1];
+	gen_rand_alphanumeric_no_underscore(dpp->get_cct(), buf, OBJ_INSTANCE_LEN);
+        block.version = buf; // using gen_rand_alphanumeric_no_underscore for the time being
+      } 
+      
+      deleteBlock = block;
+      deleteBlock.cacheObj.objName = "_:" + deleteBlock.version + "_" + deleteBlock.cacheObj.objName;
 
-  Attrs::iterator attrs;
-  Attrs currentattrs = source->get_attrs();
-  std::vector<std::string> currentFields;
-  
-  /* Extract fields from current attrs */
-  for (attrs = currentattrs.begin(); attrs != currentattrs.end(); ++attrs) {
-    currentFields.push_back(attrs->first);
-  }
+      if ((ret = blockDir->set(dpp, &deleteBlock, y)) == 0) {
+	if ((ret = blockDir->set(dpp, &block, y)) < 0) {
+	  ldpp_dout(dpp, 0) << "Failed to set head object in block directory for: " << source->get_key().get_oid() << ", ret=" << ret << dendl;
+	  return ret;
+	}
+      } else {
+	ldpp_dout(dpp, 0) << "Failed to set delete marker block in block directory for: " << source->get_key().get_oid() << ", ret=" << ret << dendl;
+	return ret;
+      }
+
+      return 0;
+    } else {
+      version = source->get_instance();
+    }
 
-  if ((ret = source->driver->get_cache_driver()->del(dpp, source->get_key().get_oid(), y)) < 0) 
-    ldpp_dout(dpp, 0) << "D4NFilterObject::" << __func__ << "(): CacheDriver del() method failed, ret=" << ret << dendl;
+    policy_prefix = head_oid_in_cache;
+    if (block.cacheObj.dirty) { // head object dirty flag represents object dirty flag
+      objDirty = true;
+      policy_prefix.erase(0, 2); // remove "D_" prefix from policy key
+    }    
 
-  return next->delete_obj(dpp, y, flags);
-}
+    if (block.deleteMarker == false) { // provided version is not a delete marker and contains data
+      block.cacheObj.objName = source->get_key().get_oid();
+      if ((ret = blockDir->del(dpp, &block, y)) == 0 || ret == -ENOENT) {
+	if ((ret = source->driver->get_cache_driver()->delete_data(dpp, head_oid_in_cache, y)) == 0) { // Sam: do we want del or delete_data here? 
+	  if (!(ret = source->driver->get_policy_driver()->get_cache_policy()->erase(dpp, policy_prefix, y))) {
+	    ldpp_dout(dpp, 0) << "Failed to delete head policy entry for: " << source->get_key().get_oid() << ", ret=" << ret << dendl;
+	    return ret;
+	  }
+	} else {
+	  ldpp_dout(dpp, 0) << "Failed to delete head object for: " << source->get_key().get_oid() << ", ret=" << ret << dendl;
+	  return ret;
+	}
+      } else {
+	ldpp_dout(dpp, 0) << "Failed to delete head object in block directory for: " << source->get_key().get_oid() << ", ret=" << ret << dendl;
+	return ret;
+      }
+      
+      if (block.cacheObj.objName != source->get_name()) {
```

Comment:
what is being done here?

Reply:
can you describe the test steps which will lead to this scenario? what is a non-versioned head object? is it the one that contains the latest version?

Reply:
The non-versioned head object is the object that contains the latest version but doesn't contain the version in the key. During a PUT in a versioned bucket, the non-versioned head object is created and set by the `set_head_obj_dir_entry` method using the object name (so the block does not use the version in the key). During a versioned delete on this same object, the head object will be left over. However, now that I'm considering this, I think the head object should be left over because other versions can still exist for that object and the head object should still exist to represent this state. I would have to update the head object to contain the correct current version after the supplied version is deleted and delete the head object if there are no more versions of the object left.